### PR TITLE
Add support for reference lookups within methods

### PR DESCRIPTION
**Pull request by JonatanWaern**
- **Add declaration spans to methods**
- **Make fields of structure statements public**
- **Fix declarations in for loops headers**
- **Do not discard if-statements without elses**
- **Allow objects_to_symbols to report errors**
- **Use try-lock to detect lookup loops**
- **Macro for internal errors**
- **Special handling of lookups on method contexts**

### CODE REVIEW

**Comment by JonatanWaern** [comment]
@TSonono Would be very useful if you could check out and build this locally to test it on your devices a bit. I will be doing some edge-case testing myself (I think, already, I can see declarations leaking through scopes in some cases)

**Comment by JonatanWaern** [comment]
Scope-escaping should be less now, though turning off reference caching for lookups within methods is certainly not ideal.

File: src/analysis/mod.rs [hunk 1]
```diff
@@ -312,6 +314,53 @@ pub struct IsolatedAnalysis {
     pub errors: Vec<DMLError>,
 }
 
+// Invariant: range covers all ranges in val
+#[derive(Debug, Clone)]
+pub struct RangeEntry {
+    range: ZeroRange,
+    symbols: HashMap<String, SymbolRef>,
+    // TODO: If this ends up not being naturally sorted by insertions,
```
[the remark this answers is not in the document]
Comment (JonatanWaern) [reply src/analysis/mod.rs]: Im fairly sure the sub_ranges are not necessarily sorted, and regardless performance could go sub-linear if we used a better structure here. Changed the todo accordingly
